File: java/PathSum.java
/**
 * Created by Robbert on 1/23/15.
 * Path Sum
 * https://oj.leetcode.com/problems/path-sum/
 */
public class PathSum {
    public boolean hasPathSum(TreeNode root, int sum) {
        if (root == null) {
            return false;
        }

        if (root.left == null && root.right == null && root.val == sum) {
            return true;
        }

        return hasPathSum(root.left, sum - root.val)
                || hasPathSum(root.right, sum - root.val);
    }
}

File: python/path_sum.py
#!/usr/bin/env python

# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @param sum, an integer
    # @return a boolean
    def hasPathSum(self, root, sum):
        if not root:
            return False

        # only check the result when we hit the leaf
        if not root.left and not root.right and root.val == sum:
            return True

        # otherwise, traverse the tree from left and right
        return self.hasPathSum(root.left, sum-root.val) or \
               self.hasPathSum(root.right, sum-root.val)
